exp/services/ledgerexporter/internal/resumablemanager.go

```
@@ -0,0 +1,121 @@
+package ledgerexporter
+
+import (
+	"context"
+	"sort"
+
+	"github.com/stellar/go/historyarchive"
+)
+
+type ResumableManager interface {
+	// Find the closest ledger number to requested start but not greater which
+	// does not exist on datastore yet.
+	//
+	// start - start search from this ledger
+	// end   - stop search at this ledger.
+	//
+	// If end=0, meaning unbounded, this will substitute an effective end value of the
+	// most recent archived ledger number.
+	//
+	// return:
+	// resumableLedger - if > 0, will be the next ledger that is not populated on data store.
+	// dataStoreComplete - if true, there was no gaps on data store for bounded range requested
+	//
+	// if resumableLedger is 0 and dataStoreComplete is false, no resumability was applicable.
+	FindStart(ctx context.Context, start, end uint32) (resumableLedger uint32, dataStoreComplete bool)
```

I think `FindStart()` should return an error instead of logging the error and swallowing it. That will give the caller of `FindStart()` more flexibility on how it can handle the error

---

the intent was to have resumability be best effort and seamless to caller by not propagating the error. seems like only other action to take with error is to stop the export, which I thought was too much interruption, do we want to trigger that now? otherwise `0` allows caller to proceed with original requested export as-is.

---

> seems like only other action to take with error is to stop the export, which I thought was too much interruption, do we want to trigger that now?

I think that could be valid option. By returning the error it gives the option to the caller and the caller can decide to abort or just start from the beginning. The caller can also keep track of how often certain errors occur.

Ledger exporter may want to vary its behavior depending on how big the ledger range is. For a full history data lake, it could be better to just abort ledger exporter which will be an immediate signal to the operator that something is wrong and the issue should be investigated. Whereas if ledger exporter swallowed the error and just started ingesting from the genesis ledger, the operator might be surprised  by that behavior.

Then there is also the use case of the GCS ledger backend which is currently in progress in https://github.com/stellar/go/pull/5260 . The ledger backend interface defines the following function:

```go
GetLatestLedgerSequence(ctx context.Context) (uint32, error)
```

So there is the expectation there that errors are propagated rather than swallowed.